raft: assert on HardState field modifications

nvb opened the pull request:
Informs #125352.

This commit adds setters for the HardState fields (Term, Vote, lead, leadEpoch), allowing us to assert that these fields are only modified in expected ways.

Some bugs that this change could catch:
- term regressions
- vote changes
- lead changes in a single term
- leadEpoch regressions

The "lead changes in a single term" class of bugs is especially interesting, as we have found two of these in this thread: https://cockroachlabs.slack.com/archives/G01G8LK77DK/p1720274557285239. These are tracked in the following issues, which we may need to address ASAP:
* https://github.com/cockroachdb/cockroach/issues/127349
* https://github.com/cockroachdb/cockroach/issues/127348

The 2nd, 3rd, and 4th commits in the PR address these issues in raft by not assuming that a `MsgSnap` as always coming from the leader of `Message.Term`.

Release note: None

nvb committed:
raft: assert on HardState field modifications

Informs #125352.

This commit adds setters for the HardState fields (Term, Vote, lead, leadEpoch),
allowing us to assert that these fields are only modified in expected ways.

Some bugs that this change could catch:
- term regressions
- vote changes
- lead changes in a single term
- leadEpoch regressions

The "lead changes in a single term" class of bugs is especially interesting, as
we have found two of these in this thread: https://cockroachlabs.slack.com/archives/G01G8LK77DK/p1720274557285239.
These are tracked in the following issues, which we may need to address ASAP:
* https://github.com/cockroachdb/cockroach/issues/127349
* https://github.com/cockroachdb/cockroach/issues/127348

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133169)
<!-- Reviewable:end -->

nvb committed:
raft: use IsMsgFromLeader is stepFollower

Group the handling of learning about the leader in stepFollower.

Epic: None
Release note: None

nvb committed:
raft: use IsMsgFromLeader in stepCandidate

Group the handling of the candidate -> follower state transition
when hearing from the leader while a candidate.

Epic: None
Release note: None

nvb committed:
raft: don't assume MsgSnap comes from leader

We can't consider MsgSnap to be from the leader of Message.Term until we
address #127348 and #127349.

Epic: None
Release note: None

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1040,6 +1037,39 @@ func (r *raft) reset(term uint64) {
+	r.Term = term
+}
+
+func (r *raft) setVote(vote pb.PeerID) {
+	assertTrue(r.Vote == None || r.Vote == vote, "cannot change vote")
+	r.Vote = vote
+}
+
+func (r *raft) resetVote() {
+	r.Vote = None
+}
```
These two methods, if used incorrectly, can result in changing the vote: if we first `resetVote` (without bumping the term), and then `setVote` to a different value.

Should we make these methods stricter in this regard? We could reset all the things in `setTerm` if the term is bumped, and remove the other `reset` methods. Then we would have the guarantee that the `set` methods do ensure the vote/lead/etc are not changed.

The `resetLead` is probably a special case here. There is some code deliberately resetting the `lead` to nil, which introduces the vulnerability that it can be changed to a different lead. For this case, I think we should, instead of resetting the `lead = None`, remember a bit saying that we "revoked" this lead, and act on this bit accordingly.

pav-kv replied:
Something like:

```golang
func setTerm(term) {
	if term == r.Term {
		return
	}
	assert(term > r.Term)
	r.Vote = None
	r.Lead = None
	// etc
}

func setVote(id) {
	if id == r.Vote {
		return
	}
	assert(r.Vote == None)
	r.Vote = id
}

func resetLead() {
	assert(r.Lead != None)
	r.leadRevoked = true
}

...
```

nvanbenschoten replied:
Good points. I've addressed this as you suggested.

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1040,6 +1034,51 @@ func (r *raft) reset(term uint64) {
+}
+
+func (r *raft) resetLeadEpoch() {
+	r.leadEpoch = 0
```
Non-actionable, but just curious about the fortification protocol: is it safe to reset `leadEpoch` to 0 in the de-fortify case?

Can the `leadEpoch` be set to non-zero again during the course of the term? If yes, can it be set to a value below the last non-zero `leadEpoch`?

nvanbenschoten replied:
It is safe to reset to 0 when if either (1) the leader has told us to defortify, or (2) we have withdrawn support for the lead epoch in the store liveness network.

`leadEpoch` can then be set to a non-zero value again if a new `MsgForgetLeader` is received, which is by design. However, the value must be equal or greater than it was before it was reset to 0, because it comes from a call to `storeLiveness.SupportFor`, which is guaranteed to be monotonic.

nvanbenschoten commented on pkg/raft/raft.go:
```diff
@@ -2098,25 +2102,12 @@ func stepFollower(r *raft, m pb.Message) error {
 		r.send(m)
 	case pb.MsgApp:
-		r.electionElapsed = 0
-		// TODO(arul): Once r.lead != None, we shouldn't need to update r.lead
```
Both of the remaining callers of `resetLead` have a TODO on them to be removed. The first is in `becomePreCandidate` and the second is in the `MsgForgetLeader` handler. I agree about the existence of `resetLead` being ugly and something that would be good to eliminate.

nvanbenschoten commented on pkg/raft/raft.go:
```diff
@@ -2076,8 +2068,6 @@ func stepCandidate(r *raft, m pb.Message) error {
 			// m.Term > r.Term; reuse r.Term
 			r.becomeFollower(r.Term, r.lead)
 		}
-	case pb.MsgTimeoutNow:
```
It is removed because `MsgTimeoutNow` is caught by `IsMsgFromLeader`. With this change, a candidate will now handle the `MsgTimeoutNow` as an instruction to force campaign at the next term, instead of ignoring it.

nvanbenschoten replied:
Yes, this is new behavior. But it's desirable behavior. I've extracted it into https://github.com/cockroachdb/cockroach/pull/133490.

> Also, do we need to assume MsgTimeoutNow came from the leader? It's a forceful campaign, and in theory any node could randomly decide to do that without the leader's involvement. Perhaps the uncertainty comes from the fact that it's used both for forced campaigns and leader transfers (and in the latter case it makes sense to say that the message is made on behalf of this leader).

It used to be disruptive but safe to allow any node to decide to perform a force campaign. With leader fortification and leader leases, it is now very unsafe. A force campaign will instruct any voter to ignore the fortification promise that it has made to the leader, which could cause a lease expiration regression (overlapping leases). Force campaigns are only safe if the leader is the one initiating them, because it will step down to a follower at the same time.

> used both for forced campaigns and leader transfers

The notion of a "forced campaign", separate from a leader transfer is relatively new. It came from https://github.com/cockroachdb/cockroach/pull/104969 and I'd like to eliminate the concept with something like https://github.com/cockroachdb/cockroach/pull/133308.

nvanbenschoten commented on pkg/raft/raft.go:
```diff
@@ -2034,6 +2034,10 @@ func stepLeader(r *raft, m pb.Message) error {
 // whether they respond to MsgVoteResp or MsgPreVoteResp.
 func stepCandidate(r *raft, m pb.Message) error {
+	if IsMsgFromLeader(m.Type) {
+		r.becomeFollower(m.Term, m.From) // always m.Term == r.Term
```
Done.

nvanbenschoten commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -2615,9 +2615,14 @@ func (r *Replica) campaignLocked(ctx context.Context) {
+func (r *Replica) forceCampaignLocked(ctx context.Context, lead raftpb.PeerID) {
 	log.VEventf(ctx, 3, "force campaigning")
-	msg := raftpb.Message{To: raftpb.PeerID(r.replicaID), Type: raftpb.MsgTimeoutNow}
+	msg := raftpb.Message{
+		To: raftpb.PeerID(r.replicaID),
+		// We pretend that the message was sent from the leader.
+		From: lead,
+		Type: raftpb.MsgTimeoutNow,
+	}
```
> Is this related to the change in this commit, or a drive-by? I see that From is used in the handler for logging, is that why?

It's related to the change and to our thread above. The use of `MsgTimeoutNow` like this is a hack and is hardly safe, it's only maybe ok because of when this code is run (see `shouldCampaignAfterConfChange`). I'd like to get rid of this, but in the meantime, I'm ok making it even more clear that cockroach is lying to raft about this message coming from the leader, which in some ways it is because the leader is the one committing the entry which demotes itself to a learner.

> Also, is it safe/canonical to communicate a leader without communicating the term? I imagine if a follower receives a MsgTimeoutNow for a long gone term, it should ignore.

Good point. I added the `Term`, along with a better comment.

nvanbenschoten commented:
@pav-kv PTAL at https://github.com/cockroachdb/cockroach/pull/133490 and then give this PR one more pass when you get a change. Thanks.

nvanbenschoten commented:
@pav-kv now that https://github.com/cockroachdb/cockroach/pull/133490 merged, this has now been rebased on the latest master.

nvanbenschoten commented:
TFTR!

bors r+

nvanbenschoten commented:
blathers backport 24.3
